ciri/consensus/snapshot/local_snapshots/pruning_service.c

```diff
@@ -0,0 +1,333 @@
+                                                              .transactions_to_prune = transactions_to_prune};
+  bool entry_point_has_no_solid_entry_points_in_past_cone;
+  hash243_set_entry_t *iter = NULL, *tmp = NULL;
+  bool spent;
+
+  params.current_snapshot_index = ps->last_pruned_snapshot_index + 1;
+
+  ERR_BIND_GOTO(tangle_traversal_dfs_to_past(params.tangle, collect_transactions_for_pruning_do_func, entry_point_hash,
+                                             ps->conf->genesis_hash, NULL, &params),
+                err, cleanup);
+
+  entry_point_has_no_solid_entry_points_in_past_cone = true;
+  iter = NULL;
+  tmp = NULL;
+
+  HASH_ITER(hh, *params.transactions_to_prune, iter, tmp) {
+    if (iota_snapshot_has_solid_entry_point(ps->new_snapshot, iter->hash)) {
+      entry_point_has_no_solid_entry_points_in_past_cone = false;
+      break;
+    }
+  }
+
+  if (entry_point_has_no_solid_entry_points_in_past_cone) {
+    HASH_ITER(hh, *params.transactions_to_prune, iter, tmp) {
+      hash_pack_reset(&tx_pack);
+      ERR_BIND_GOTO(
+          iota_tangle_transaction_load_partial(tangle, iter->hash, &tx_pack, PARTIAL_TX_MODEL_ESSENCE_METADATA), err,
+          cleanup);
+      ERR_BIND_GOTO(iota_spent_addresses_service_was_tx_spent_from(sap, tangle, &tx, iter->hash, &spent), err, cleanup);
+      ERR_BIND_GOTO(tips_cache_remove(ps->tips_cache, iter->hash), err, cleanup);
+    }
+  }
+
+  if (!entry_point_has_no_solid_entry_points_in_past_cone) {
+    hash243_set_free(params.transactions_to_prune);
+  }
+
+cleanup:
+  if (err != RC_OK) {
+    params.current_snapshot_index = ps->last_pruned_snapshot_index;
+    hash243_set_free(params.transactions_to_prune);
+  }
+  return RC_OK;
+}
+
+static retcode_t prune_transactions(pruning_service_t *const ps, tangle_t const *const tangle,
+                                    spent_addresses_provider_t *const sap) {
+  retcode_t err;
+  DECLARE_PACK_SINGLE_MILESTONE(milestone, milestone_ptr, milestone_pack);
+  hash243_set_t transactions_to_prune = NULL;
+
+  ERR_BIND_GOTO(iota_tangle_milestone_load_by_index(tangle, ps->last_pruned_snapshot_index, &milestone_pack), err,
+                cleanup);
+
+  ERR_BIND_GOTO(collect_transactions_to_prune(ps, tangle, sap, milestone.hash, &transactions_to_prune), err, cleanup);
+  ERR_BIND_GOTO(iota_tangle_transaction_delete(tangle, transactions_to_prune), err, cleanup);
+  hash243_set_free(&transactions_to_prune);
+  // It's important to delete the milestone only after all it's past cone has been deleted to avoid dangle transactions
+  hash243_set_add(&transactions_to_prune, milestone.hash);
+  ERR_BIND_GOTO(iota_tangle_milestone_delete(tangle, milestone.hash), err, cleanup);
+  log_info(logger_id,
+           "Snapshot index % " PRIu64 " was pruned successfully, % " PRIu64 " transactions were removed from db\n",
+           ps->last_pruned_snapshot_index, hash243_set_size(transactions_to_prune));
```

Comment:
You won't have the right size since you freed the set. You'll always have `1` here.